### DISCUSSION TITLE
Native way to right align full widget etc. (not text) + few questions

### DISCUSSION THREAD

**Discussion by Redflashx12** [fltk-rs/fltk-rs#1464] [Q&A]
Hello,

I wanted to ask if there is a supported, native way to align a full button, frame, or any widget on the right side, e.g. when using Flex (without having to add a "spacer" frame since that seems a bit unintuitive or more like a workaround than really intended), but maybe also for Pack or other groups?

My goal is to have a play/pause button aligned on the left side and some other UI elements aligned on the right side.

In a test, I already made it work like this, but I don't feel like this is how it should work and the left/right/below/above functions do not work with Flex, right? Is there a reason for that?

```
use fltk::{prelude::*, *};
use fltk::enums::Align;
use fltk::frame::Frame;

fn main() {

    let a = app::App::default().with_scheme(app::Scheme::Gtk);
    let mut win = window::Window::default().with_size(640, 480);
    let mut col = group::Flex::default().with_size(win.width(), 40).row();

    let mut button_left = button::Button::default().with_size(120, 40).with_label("Left Side!");
    col.fixed(&button_left, 120);

    let mut spacer = Frame::default().with_size(120, 40);

    let mut button_right = button::Button::default().with_size(120, 40).with_label("Right Side!");
    col.fixed(&button_right, 120);

    let button_left_of_right = button::Button::default().with_size(120, 40).with_label("Left of Right!").left_of(&button_right, 10);
    col.fixed(&button_left_of_right, 120);

    col.set_margins(10, 10, 10, 10);
    col.set_spacing(10);

    col.end();
    win.end();
    win.show();
    win.size_range(600, 400, 0, 0);
    a.run().unwrap();
}
```

And it is not possible to fix the height in a Flex column for just a single element among many (or width for Flex row), right? You would need to create a separate group just for this one element in order to do that, right?

And is there a nice overview for all the different adjustable parameters like margins, padding, spacing and so on and what it does exactly? I thought it followed the CSS box model but 'spacing' is not part of the Box model and if I tested it correctly, it creates space among the elements in the group itself and the margin is only referring to the next group, is that correct? And what is the difference between padding and spacing? When I try to exchange pad with spacing there at the end (where col.set_margins/spacing is located) there's no difference in the UI.

**Comment by MoAlyousef** [comment]
Hello

For elements inside a Flex, specifying any sizes with with_size() or in new() won't work, you only need to specify it using the flex's fixed() method.
So in your code example, it would be:
```rust
use fltk::{prelude::*, *};
use fltk::enums::Align;
use fltk::frame::Frame;

fn main() {

    let a = app::App::default().with_scheme(app::Scheme::Gtk);
    let mut win = window::Window::default().with_size(640, 480);
    let mut col = group::Flex::default().with_size(win.width(), 40).row();

    let mut button_left = button::Button::default().with_label("Left Side!");
    col.fixed(&button_left, 120);

    let mut spacer = Frame::default();

    let button_left_of_right = button::Button::default().with_label("Left of Right!");
    col.fixed(&button_left_of_right, 120);

    let mut button_right = button::Button::default().with_label("Right Side!");
    col.fixed(&button_right, 120);

    col.set_margins(10, 10, 10, 10);
    col.set_spacing(10);

    col.end();
    win.end();
    win.show();
    win.size_range(600, 400, 0, 0);
    a.run().unwrap();
}
```
Also the call to left_of will be ignored by the flex, since it does it's layout based on the order of elements. So button_left_of_right actually appears on the right of button_right since it's inserted after it. So I placed it before button_right here.

There's a [fltk-grid]() crate which might allow you to position widgets in a grid-like manner.
https://github.com/fltk-rs/fltk-grid

Otherwise FLTK will be adding a Grid widget which I intend to wrap as soon as it lands in master.

> And it is not possible to fix the height in a Flex column for just a single element among many (or width for Flex row), right? You would need to create a separate group just for this one element in order to do that, right?

That's what fixed is for. You shouldn't need to create a separate group. If you have a code example to show what you mean that would be helpful.

Regarding the terms used with Flex:
spacing and pad are the same. They define the spacing between widgets.

The reason is that before FLTK had a Flex widget, fltk-rs used a custom one based on the osen/Fl_Flex library. Once FLTK added a Flex widget, we shifted to that.

set_margin defines the all round margin around the whole Flex, while set_margins allows you to individually specify the margins to left, top, right and bottom.
